### PR TITLE
Frontend arch feature

### PR DESCRIPTION
**Pull request by MCGallaspy**
## Summary

Supersedes https://github.com/learningequality/kolibri/pull/82.

@dxcanas @indirectlylit lit take note -- after this is merged into master, any open PRs targeting `frontend-arch-base` should be rebased against master and target that branch now instead. We can then phase out the `frontend-arch-feature` branch for now.

## TODO

- [X] ~~Have tests been written for the new code?~~ Maybe not enough though?
- [X] Has documentation been written/updated?
- [X] New dependencies (if any) added to requirements file
- [ ] Add an entry to CHANGELOG.rst

## Reviewer guidance

Sadly I have no words of wisdom. :(

### CODE REVIEW

File: kolibri/core/templates/kolibri/base.html [hunk 1]
```diff
@@ -18,10 +15,11 @@
 </ul>
 {% endblock %}
 
-<hr>
-
-{% block content %}
-
+<main></main>
+{% block frontend_assets %}
```
Comment (rtibbles) [inline kolibri/core/templates/kolibri/base.html:19]: Why is this moved into the body of the html? This block should be in the `<head>`.

File: kolibri/plugins/audio_mp3_render/kolibri_plugin.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import logging
+
+from kolibri.plugins.base import KolibriFrontEndPluginBase
+from kolibri.plugins.hooks import CONTENT_RENDERER_ASYNC
+
+logger = logging.getLogger(__name__)
+
+
+class KolibriAudioMP3RenderFrontEnd(KolibriFrontEndPluginBase):
```
Comment (rtibbles) [inline kolibri/plugins/audio_mp3_render/kolibri_plugin.py]: This has not been migrated to the new hooks.

File: kolibri/plugins/document_pdf_render/kolibri_plugin.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+import logging
+
+from kolibri.plugins.base import KolibriFrontEndPluginBase
+from kolibri.plugins.hooks import CONTENT_RENDERER_ASYNC
+
+logger = logging.getLogger(__name__)
+
+
+class KolibriDocumentPDFRenderFrontEnd(KolibriFrontEndPluginBase):
```
Comment (rtibbles) [inline kolibri/plugins/document_pdf_render/kolibri_plugin.py]: This has not been migrated to the new hooks.

File: kolibri/plugins/video_mp4_render/kolibri_plugin.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+import logging
+
+from kolibri.plugins.base import KolibriFrontEndPluginBase
+from kolibri.plugins.hooks import CONTENT_RENDERER_ASYNC
+
+logger = logging.getLogger(__name__)
+
+
+class KolibriVideoMP4RenderFrontEnd(KolibriFrontEndPluginBase):
```
Comment (rtibbles) [inline kolibri/plugins/video_mp4_render/kolibri_plugin.py]: This has not been migrated to the new hooks.

File: kolibri/core/templates/kolibri/base.html (hunk 1 above)
Comment (MCGallaspy) [inline kolibri/core/templates/kolibri/base.html:19]: The ready method of the learn KolibriModule must be executed _after_ the `<main>` tag is created... which is not the case if that code is included in the `<head>`

File: kolibri/plugins/audio_mp3_render/kolibri_plugin.py [hunk 5]
```diff
@@ -1,33 +0,0 @@
...
-
-import logging
-
-from kolibri.plugins.base import KolibriFrontEndPluginBase
-from kolibri.plugins.hooks import CONTENT_RENDERER_ASYNC
-
-logger = logging.getLogger(__name__)
-
-
-class KolibriAudioMP3RenderFrontEnd(KolibriFrontEndPluginBase):
```
Comment (rtibbles) [inline kolibri/plugins/audio_mp3_render/kolibri_plugin.py]: I would rather these were migrated to the new system, rather than just deleted. They do serve a purpose.

File: kolibri/core/templates/kolibri/base.html (hunk 1 above)
Comment (rtibbles) [inline kolibri/core/templates/kolibri/base.html:19]: We should hold off any rendering through Vue if the document is not ready yet, not put our script tags in the middle of our HTML.

http://stackoverflow.com/questions/436411/where-is-the-best-place-to-put-script-tags-in-html-markup

We can add the async directives later, if we want to optimize page load, but this functionality should be implemented by listening for a documentContentLoaded event.

Comment (rtibbles) [inline kolibri/core/templates/kolibri/base.html:19]: (Note, we should probably only implement this once, in the global Vue object, and then only allow things to render when this has come to pass).

Comment (indirectlylit) [inline kolibri/core/templates/kolibri/base.html:19]: There isn't any HTML to parse since the UI is completely dynamically generated, so in practical terms shouldn't be any performance difference. This just seemed easier.

Comment (MCGallaspy) [inline kolibri/core/templates/kolibri/base.html:19]: Well, to be honest I expected the `ready` method to execute after the page has loaded. But it does not. Does it have a well-defined time when it's executed? Can we just make _that_ do its thing after page load?

Comment (rtibbles) [inline kolibri/core/templates/kolibri/base.html:19]: The ready method is fired whenever the module has been registered. It doesn't know anything about HTML rendering, that's deferred to Vue.

Comment (MCGallaspy) [inline kolibri/core/templates/kolibri/base.html:19]: Okay, so any DOM manipulations in the ready method should be deferred until after page load? We can do that.
